### ISSUE TITLE
Training with autobatching seems to be slower than before

### ISSUE THREAD

**Issue by shuheik** [clab/dynet#974]
I suspect that training with autobatching got a bit slower after some recent commit. 
I compared some examples (train_rnn-autobatch in examples, bilstm-tagger-withchar-bulk, treenn-bulk and rnnlm-batch in neulab/dynet-benchmark) between the latest (a69c515) and an old (a3782a1) commit, and found that old ones were 1.4 ~ 1.8 times faster in all of them. 
I chose a3782a1 for comparison because I remembered it was faster, and I don't think it is _the_ last commit without this speed issue.

**Comment by shuheik** [comment]
Here are some actual numbers on a GPU machine with GTX 1070.

train_rnn-autobatch with auto batching 
latest: sent/sec: 280.034
old: sent/sec: 397.141

train_rnn-autobatch with manual batching
latest: sent/sec: 766.284
old: sent/sec: 747.943

The numbers below are the first ones reported by the program, but the trend seems stable.

bilstm-tagger-withchar-bulk-gpu --dynet-autobatch 1 20 128 50 32 0 32 2 600
latest: word_per_sec=2540.74
old: word_per_sec=4680.32

treenn-bulk-gpu --dynet-autobatch 1 --dynet-mem 4096 128 128 0 32 2 600
latest: sent_per_sec=1001.88
old: sent_per_sec=1420.21

rnnlm-batch-gpu 64 128 256 0 600
latest: word_per_sec=45789.7
old: word_per_sec=48097.8

**Comment by neubig** [comment]
Hmm this is a bad regression. We'll take a look.

**Comment by pmichel31415** [comment]
Are there any news on that? I'm delaying updating dynet because I have models that rely heavily on autobatching.

**Comment by neubig** [comment]
@pmichel31415 No, sorry. One thing that would be really helpful is if someone could do a binary search through commit space to pinpoint the commit where performance regressed. I think this could probably be automated so effort would be minimal, but I don't have time to do it right now...

**Comment by neubig** [comment]
Thanks, this is super-helpful. I'll try to fix this ASAP.

**Comment by neubig** [comment]
OK, so there's been some progress in #1029. To summarize there was one really bad regression related to Eigen streams, but there's also something else so we're not back up to the original speed:

* Before Eigen stream modification: word_per_sec=16736.5
* After Eigen stream modification: word_per_sec=7432.61
* Latest master: word_per_sec=4292.2
* This PR: word_per_sec=11800.2

I'll continue looking for the second part of the regression.

**Comment by neubig** [comment]
OK, it looks like the other regression was in `d9f110a090cc39396fb72e65549160443e08fc94`, the uber-commit that added broadcasting. This commit touches a ton of code, so we'll have to do some profiling to figure out exactly what is causing the problem. Here is a minimal pair of commits that can reproduce the problem:

* `b9cae3446c79e921b85072e2a46390149d5fc9e8`: word_per_sec=7259.31
* `d9f110a090cc39396fb72e65549160443e08fc94`: word_per_sec=4287.64

This is using the following command in the `dynet-cpp` folder of https://github.com/neulab/dynet-benchmark/:

    ./bilstm-tagger-withchar-bulk --dynet-autobatch 1 20 128 50 32 0 32 2 600

@msperber, this is your commit so if you could take a look that'd be great, but I'll also try to debug myself.

**Comment by msperber** [comment]
I'll try to take a look also. Was this CPU or GPU speed?

**Comment by neubig** [comment]
GPU, thanks!

**Comment by neubig** [comment]
@msperber A preliminary investigation seems to suggest that both the forward and backward steps of cmult are slow. I wonder if this is due to broadcasting over 4 dimensions instead of not broadcasting at all, or broadcasting over only a single dimension.

**Comment by msperber** [comment]
I noticed that the same_dims check, which was meant to bypass broadcast if not needed, was a bit leaky, which might have caused the issue. The benchmarks you were trying didn't actually use any broadcasting, right? I've opened a WIP PR, maybe that already helps. We could also try to catch the case where only the left or right side (not both) need broadcasting, which is probably almost always the case, but that wouldn't fix the case of not using any broadcasting, of course.
